metricflow/dataflow/optimizer/source_scan/cm_branch_combiner.py

```diff
@@ -0,0 +1,437 @@
+            )
+            return ComputeMetricsBranchCombinerResult[SourceDataSetT](combined_node)
+
+        self._log_combine_failure(
+            left_node=self._current_left_node,
+            right_node=current_right_node,
+            combine_failure_reason="there are functional differences",
+        )
+        return ComputeMetricsBranchCombinerResult()
+
+    def visit_source_node(  # noqa: D
+        self, node: ReadSqlSourceNode[SourceDataSetT]
+    ) -> ComputeMetricsBranchCombinerResult:
+        self._log_visit_node_type(node)
+        return self._default_handler(node)
+
+    def visit_join_to_base_output_node(  # noqa: D
+        self, node: JoinToBaseOutputNode[SourceDataSetT]
+    ) -> ComputeMetricsBranchCombinerResult:
+        self._log_visit_node_type(node)
+        return self._default_handler(node)
+
+    def visit_join_aggregated_measures_by_groupby_columns_node(  # noqa: D
+        self, node: JoinAggregatedMeasuresByGroupByColumnsNode[SourceDataSetT]
+    ) -> ComputeMetricsBranchCombinerResult:
+        self._log_visit_node_type(node)
+        return self._default_handler(node)
+
+    def visit_aggregate_measures_node(  # noqa: D
+        self, node: AggregateMeasuresNode[SourceDataSetT]
+    ) -> ComputeMetricsBranchCombinerResult:
+        self._log_visit_node_type(node)
+        current_right_node = node
+
+        combined_parent_nodes = self._combine_parent_branches(current_right_node)
+        if combined_parent_nodes is None:
+            return ComputeMetricsBranchCombinerResult()
+
+        if not isinstance(self._current_left_node, current_right_node.__class__):
+            self._log_combine_failure(
+                left_node=self._current_left_node,
+                right_node=current_right_node,
+                combine_failure_reason=f"left node is not a {current_right_node.__class__.__name__}",
+            )
+            return ComputeMetricsBranchCombinerResult()
+
+        assert len(combined_parent_nodes) == 1
+        combined_parent_node = combined_parent_nodes[0]
+        assert combined_parent_node is not None
+
+        combined_metric_input_measure_specs = (
+            self._current_left_node.metric_input_measure_specs + current_right_node.metric_input_measure_specs
+        )
+
+        for spec in combined_metric_input_measure_specs:
+            # Avoid combining branches if the AggregateMeasuresNode specifies a metric with an alias to avoid
+            # collisions e.g. two metrics use the same alias for two different measures. This is not always the case,
+            # so this could be improved later.
+            if spec.alias is not None:
+                self._log_combine_failure(
+                    left_node=self._current_left_node,
+                    right_node=current_right_node,
+                    combine_failure_reason=f"Metric input measure spec {spec} has an alias",
+                )
+                return ComputeMetricsBranchCombinerResult()
```

Comment:
This is all very difficult to reason about. In theory this shouldn't be an issue if we take the distinct set of combined input measure specs, but I guess if we combine parents in a way that assembles two different measures (or metrics, when we do this for derived metrics) that happen to have the same alias that'll cause problems.

We could make this less aggressive by updating the conditional to be something like:

```
aliases = [spec.alias for spec in combined_metric_input_measure_specs]
distinct_aliases = set(aliases)
if len(aliases) > len(distinct_aliases):
    return ...
```